doc/metadata-server/README.md

```diff
@@ -0,0 +1,327 @@
+
+- Instance metadata (`InsMetadata`) - part of the metadata dedicated related only to the `Instance`. For example we can imagine that the user picks a `ExtContract` from the `Marketplace` and in `Marlowe Run` he is be able to add its own information set to this particular instance (add own description, describe role differently, change title etc.). Parts of this metadata set could be probably complately service dependent (for example in different instances / deployments of `Marlowe Run` we can use different schema here and allow attachment of extra documents).
+
+- Marlowe contract metadata (`ContractMetadata`) - the same as in the previous section but now we can look at it as a product: `(InsMetadata, ExtMetadata)`.
+
+# Verification of `ContractMetadata`
+
+In the current scenario we want to verify the validity of two parts of `ContractMetadata` (`(InsMetadata, ExtMetadata)`). To do this we can slightly modify the previous verification schema. `InsMetadata` can be verified using the same strategy as before because it is information dedicated for a specific Contract so we need to prove the connection of the metadata with onchain triple: roles distribution, `Validator` and `Contract`.
+
+When we want to verify the `ExtContract` and `ExtMetadata` are really associated with the data on-chain we should prove that there is a set of parameters which when applied to `ExtContract` produced the same `Contract` value as we have in this particular `Tx`. `ExtContract` and `ExtMetadata` association should be proved by a separate signature / certificate.
+
+<!-- IMAGE Extended contract comparision flow. -->
+
+```purescript
+type RoleDistribution = Map Role PubKeyHash
+
+type Params = Map Id Marlowe.Semantics.Value
+
+type MarloweTxMetadata =
+  { instance ::
+    { -- | Here we have the same set of data as before
+    }
+  , template ::
+    { extContract :: ExtContract
+    , params :: Params
+    , metadata :: ExtMetadata
+    -- | Prove that `metadata` set is dedicated for the
+    -- | above `extContract` value.
+    , signature :: Signature
+    }
+  }
+
+type Metadata = { instance :: ..., template :: ... }
+
+extractMetadata :: OnChainTx -> Maybe Metadata
+extractMetadata tx = do
+  marloweTxMetadata <- decode <$> extractMetadata tx
+  txContract <- extractMarloweContract tx
+
+  let
+    templateTxMetadata = marloweTxMetadata.template
+
+  extMetadata <- do
+    if fillTemplate templateTxMetadata.extContract templateTxMetadata.params == txContract
+      && checkValidator txValidator
+      && checkSignature (templateTxMetadata.extContract /\ templateTxMetadata.metadata) MarloweTxMetadata.signature
+        then
+          Just templateTxMetadata.metadata
+        else
+          Nothing
+
+  -- | For the `InsMetadata` we reuse a similar procedure
+  -- | as in previous chapter
+  insMetadata <- extractInstanceMetadata
+
+  pure { insMetadata, extMetadata }
+```
+
+## Protocol
```

Comment:
This sounds like it could be more context if it was at the beginning, before the schema. But it could use a bit more introduction/context. What does the solution look like?